### PR TITLE
 feat: restrict login to users matching a certain group

### LINKED ISSUE #867: Restrict login to users matching a certain group
We would like to limit access to Nextcloud to users having certain groups. So when a user authenticates with OAuth, it should check the groups coming from the OAuth provider before a user is created / allowed to login in. If the user is not on a whitelist (regex), the user login is rejected.

## Background

We have a member database containing many members from different divisions. Some divisions have their own Nextcloud instance, and it would be great to allow them to use the member database for login. But other divisions should not be able to use the Nextcloud instance. The user story is described in https://github.com/hitobito/hitobito_jubla/issues/74 (german)

## Proposal

- Add a new field in the settings for setting a whitelist regex for which groups (gid) are synchronized with Nextcloud
- Add a new checkbox in the settings "Restrict login for users without whitelisted groups"
- During the authentication process, check if the user is in at least one group that is part of the whitelist
- During the group synchronisation, only add and remove groups that are whitelisted. (this would also solve #866)

What do you think about this proposal?

### PR DESCRIPTION
**Pull request by bergerar**
Fixes  #867

### CODE REVIEW

**Comment by edward-ly** [comment]
Closing as discussed in https://github.com/nextcloud/user_oidc/issues/867#issuecomment-2406113586.

**Comment by Stefanqn** [comment]
I'm quite surprised about this decision: the referenced discussion describes a cumbersome workaround, including creating custom Keycloak flows and carefully adding the Keycloak plugin to multiple locations in these flows. Furthermore, it's an unofficial, single-developer plugin and it limits the solution space to Keycloak. On the other hand, a pull request with a proper solution at PEP (Policy Enforcement Point) exists and gets rejected. Would you please elaborate on this decision?

**Comment by edward-ly** [comment]
We happily welcome open discussion from anyone, not just those already mentioned. We're not saying that this feature isn't useful, but rather that it's better to implement it as a separate app or use LDAP as an alternative.

I agree that the Keycloak extension isn't a perfect solution, but what about my point above? What does this PR provide that another app doesn't (user_ldap or otherwise)?

**Comment by Stefanqn** [comment]
Thank you Edward for keeping an open mind on this. 

Regarding your point(s):
a)  use LDAP as an alternative: I'm quite confused about this suggestion as it would result in switching to a completely different protocol for authorization - one many don't have in place. Maybe you can describe your idea in more detail.
b)  implement it as a separate app: what would be the motivation to do so? You mentioned user_LDAP directly implementing this, so why do it differently for user_oidc?

**Comment by edward-ly** [comment]
Let's continue the discussion in https://github.com/nextcloud/user_oidc/issues/867#issuecomment-2435647720 in which I made a response there.

**Comment by edward-ly** [comment]
Reopening as per https://github.com/nextcloud/user_oidc/issues/867#issuecomment-2438267667, will test and merge within the next few days.

**Review by edward-ly — CHANGES_REQUESTED** [review]
Please rebase the branch to start at the latest (main) commit and resolve any conflicts before we can start testing and merging, thank you.

**Comment by bergerar** [comment]
I rebased the pull request, unit tests and lints are passing and I did some manual tests in my setup of the feature.

File: lib/Controller/LoginController.php [hunk 1]
```diff
@@ -549,6 +549,18 @@ public function code(string $state = '', string $code = '', string $scope = '',
...
 		}
 
+		// prevent login of users that are not in a whitelisted group (if activated)
+		$restrictLoginToGroups = $this->providerService->getSetting($providerId, ProviderService::SETTING_RESTRICT_LOGIN_TO_GROUPS, '0');
+		if($restrictLoginToGroups === '1') {
+			$syncGroups = $this->provisioningService->getSyncGroupsOfToken($providerId, $idTokenPayload);
+
+			if($syncGroups === null || count($syncGroups) === 0) {
+				$this->logger->debug('Prevented user from login as user is not part of a whitelisted group');
+				$message = $this->l10n->t('You do not have permission to log in to this instance. If you believe this is an error, please contact an Administrator.');
```
Comment (julien-nc) [inline lib/Controller/LoginController.php:559]:
```suggestion
				$message = $this->l10n->t('You do not have permission to log in to this instance. If you think this is an error, please contact an administrator.');
```

File: lib/Controller/LoginController.php [hunk 2]
```diff
@@ -549,6 +549,18 @@ public function code(string $state = '', string $code = '', string $scope = '',
...
+				return $this->build403TemplateResponse($message, Http::STATUS_FORBIDDEN, ['reason' => 'user not allowed to login']);
```
Comment (julien-nc) [inline lib/Controller/LoginController.php:560]:
```suggestion
				return $this->build403TemplateResponse($message, Http::STATUS_FORBIDDEN, ['reason' => 'user not in any whitelisted group']);
```

File: src/components/SettingsForm.vue [hunk 3]
```diff
@@ -257,6 +257,21 @@
...
 			{{ t('user_oidc', 'This will create and update the users groups depending on the groups claim in the id token. The Format of the groups claim value should be [{gid: "1", displayName: "group1"}, ...] or ["group1", "group2", ...]') }}
 		</p>
+		<p>
+			<label for="group-whitelist-regex">{{ t('user_oidc', 'Group whitelist regex') }}</label>
+			<input id="group-whitelist-regex"
+				v-model="localProvider.settings.groupWhitelistRegex"
+				type="text">
+		</p>
+		<p class="settings-hint">
+			{{ t('user_oidc', 'Only groups matching the whitelist regex will be created, updated and deleted by the group claim') }}
```
Comment (julien-nc) [inline src/components/SettingsForm.vue:267]:
```suggestion
			{{ t('user_oidc', 'Only groups matching the whitelist regex will be created, updated and deleted by the group claim. For example: {regex} allows all groups which ID starts with {substr}', { regex: '/^blue/', substr: 'blue' }) }}
```
Or anything more descriptive.

File: src/components/SettingsForm.vue [hunk 4]
```diff
@@ -257,6 +257,21 @@
...
+		</p>
+		<NcCheckboxRadioSwitch :checked.sync="localProvider.settings.restrictLoginToGroups" wrapper-element="div">
+			{{ t('user_oidc', 'Restrict login for users without whitelisted groups.') }}
```
Comment (julien-nc) [inline src/components/SettingsForm.vue:270]:
```suggestion
			{{ t('user_oidc', 'Restrict login for users that are not in any whitelisted group') }}
```

File: src/components/SettingsForm.vue [hunk 5]
```diff
@@ -257,6 +257,21 @@
...
+		</NcCheckboxRadioSwitch>
+		<p class="settings-hint">
+			{{ t('user_oidc', 'Users that are not part of a whitelisted group are not created and can not login') }}
```
Comment (julien-nc) [inline src/components/SettingsForm.vue:273]:
```suggestion
			{{ t('user_oidc', 'Users that are not part of any whitelisted group are not created and can not login') }}
```

**Review by julien-nc — CHANGES_REQUESTED** [review]
Looks sane to me. Tested, works fine :heavy_check_mark: 

A small improvement could be to check if the leading and trailing slashes are there and add them if not.

The labels/hints in the settings need to be as much descriptive as possible. See inline comments.

**Comment by bergerar** [comment]
@julien-nc  the delimiters are now added automatically if the first one is missing.

There was one check failing, but it was the same one as on the main branch (Integration tests / php8.1-mysql-stable27)

**Comment by bergerar** [comment]
still had an old version of the php-cs-fixer.

I did now rebase, update install the updated dependencies and run `composer run cs:fix`.

@julien-nc is should now pass all tests

**Review by julien-nc — APPROVED** [review]

**Comment by bergerar** [comment]
Thanks for merging and your work in general on open source. @julien-nc

**Comment by julien-nc** [comment]
@bergerar Thank you for your patience, your perseverance, your nice words and your contribution!

**Pull request merged**